### ISSUE TITLE
Random TCP timeouts when accessing SVCs intra-cluster using ClusterIP

### ISSUE THREAD

**Issue by harsha-y** [kubernetes/kops#4323]
We are able to reproduce this problem across multiple clusters in different environments. But for simplicity sake, the test cluster we are currently running has 3 masters and 2 nodes.

1. `kops version`
```
Version 1.8.0
```

2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.
```
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.2", GitCommit:"5fa2db2bd46ac79e5e00a4e6ed24191080aa463b", GitTreeState:"clean", BuildDate:"2018-01-18T21:12:46Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.4", GitCommit:"9befc2b8928a9426501d3bf62f72849d5cbcd5a3", GitTreeState:"clean", BuildDate:"2017-11-20T05:17:43Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```

3. What cloud provider are you using?
```
AWS
```

4. What commands did you run?  What is the simplest way to reproduce this issue?
Exec into the running pods of the test application and run curl in a loop - 
```
$ kubectl exec -it test-service-89bb6f57-5smbw sh
/opt/app # for i in $(seq 1 10); do curl --connect-timeout 1 http://test-service-sv/ping && echo ; done;
pong
curl: (28) Connection timed out after 1001 milliseconds
curl: (28) Connection timed out after 1000 milliseconds
pong
curl: (28) Connection timed out after 1001 milliseconds
curl: (28) Connection timed out after 1001 milliseconds
pong
pong
pong
curl: (28) Connection timed out after 1001 milliseconds
```

This test java application `test-service` that has the endpoint `/ping` created as a k8s `deployment`. When a get request is sent to this endpoint, it responds with a HTTP 200 and a message `pong`.

<details>
  <summary>Deployment</summary>

```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: test-service
  namespace: default
spec:
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 0
  minReadySeconds: 0
  revisionHistoryLimit: 10
  replicas: 3
  template:
    metadata:
      labels:
        name: test-service
    spec:
      containers:
        - name: test-service
          image: earnest/test-service:1.2-10fb473
          resources:
            requests:
              memory: "256Mi"
              cpu: "250m"
            limits:
              memory: "512Mi"
              cpu: "500m"
          ports:
            - name: default-port
              containerPort: 3000
          livenessProbe:
            tcpSocket:
              port: default-port
            initialDelaySeconds: 10
            failureThreshold: 24
            timeoutSeconds: 3
          readinessProbe:
            httpGet:
              path: /ping
              port: default-port
            initialDelaySeconds: 10
            failureThreshold: 3
            timeoutSeconds: 3
      imagePullSecrets:
        - name: dockerregistrykey
```
</details>

<details>
  <summary>SVC</summary>

```
apiVersion: v1
kind: Service
metadata:
  name: test-service-sv
  labels:
    name: test-service
spec:
  type: ClusterIP
  ports:
    - port: 80
      targetPort: default-port
      protocol: TCP
  selector:
    name: test-service
```
</details>

5. What happened after the commands executed?
- On nodes that have more than have 2 or more Pods of the deployment - We see timeouts, seemingly at random, when we run the above curl from inside any of the pods on these nodes
- On nodes that have 1 pod of the deployment - We see no timeouts
- We can replicate the problem by using the ClusterIP instead of the internal SkyDNS address
- We cannot replicate the problem if we use individual endpoint ip addresses instead of the ClusterIP
- We see no timeouts if we use curl loops from the nodes - this seems to indicate that the problem is restricted to k8s SVCs and the pods but not a host level network problem
- Using tcpdump we are able to verify that the SYN packets are reaching the target endpoints but the ACK packets meant for the source never reach the source. This causes TCP to go into exponential backoff mode and finally timing out at ~130 seconds.

6. What did you expect to happen?
We expect to see no timeouts

7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -oyaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.

<details>
  <summary>Cluster Manifest</summary>

```
apiVersion: kops/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: 2018-01-11T19:23:22Z
  name: kubernetes.test.io
spec:
  api:
    loadBalancer:
      type: Internal
  authorization:
    rbac: {}
  channel: stable
  cloudLabels:
    KubernetesCluster: kubernetes.test.io
    created_by: INF
  cloudProvider: aws
  configBase: s3://kubernetes-test-io-state-store/kubernetes.test.io
  etcdClusters:
  - etcdMembers:
    - instanceGroup: master-us-east-1a
      name: a
    - instanceGroup: master-us-east-1b
      name: b
    - instanceGroup: master-us-east-1c
      name: c
    name: main
  - etcdMembers:
    - instanceGroup: master-us-east-1a
      name: a
    - instanceGroup: master-us-east-1b
      name: b
    - instanceGroup: master-us-east-1c
      name: c
    name: events
  iam:
    allowContainerRegistry: true
    legacy: false
  kubernetesApiAccess:
  - 0.0.0.0/0
  kubernetesVersion: 1.8.4
  masterInternalName: api.internal.kubernetes.test.io
  masterPublicName: api.kubernetes.test.io
  networkCIDR: 172.30.0.0/16
  networkID: vpc-c6e25fa2
  networking:
    flannel:
      backend: vxlan
  nonMasqueradeCIDR: 100.64.0.0/10
  sshAccess:
  - 0.0.0.0/0
  subnets:
  - cidr: 172.30.1.0/24
    egress: nat-0647893be0d746e46
    id: subnet-dc714ff7
    name: us-east-1a
    type: Private
    zone: us-east-1a
  - cidr: 172.30.3.0/24
    egress: nat-0647893be0d746e46
    id: subnet-c10efbb7
    name: us-east-1b
    type: Private
    zone: us-east-1b
  - cidr: 172.30.5.0/24
    id: subnet-75d9d62c
    name: us-east-1c
    type: Private
    zone: us-east-1c
  topology:
    dns:
      type: Public
    masters: private
    nodes: private

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2018-01-11T19:23:22Z
  labels:
    kops.k8s.io/cluster: kubernetes.test.io
  name: master-us-east-1a
spec:
  associatePublicIp: false
  image: ami-e582d29f
  machineType: m4.large
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-us-east-1a
  role: Master
  subnets:
  - us-east-1a

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2018-01-11T19:23:23Z
  labels:
    kops.k8s.io/cluster: kubernetes.test.io
  name: master-us-east-1b
spec:
  associatePublicIp: false
  image: ami-e582d29f
  machineType: m4.large
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-us-east-1b
  role: Master
  subnets:
  - us-east-1b

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2018-01-11T19:23:23Z
  labels:
    kops.k8s.io/cluster: kubernetes.test.io
  name: master-us-east-1c
spec:
  associatePublicIp: false
  image: ami-e582d29f
  machineType: m4.large
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-us-east-1c
  role: Master
  subnets:
  - us-east-1c

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2018-01-11T19:23:23Z
  labels:
    kops.k8s.io/cluster: kubernetes.test.io
  name: nodes
spec:
  additionalSecurityGroups:
  - sg-018ff173
  associatePublicIp: false
  image: ami-e582d29f
  machineType: m4.large
  maxSize: 2
  minSize: 2
  nodeLabels:
    kops.k8s.io/instancegroup: nodes
  role: Node
  subnets:
  - us-east-1a
  - us-east-1b
  - us-east-1c
```
</details>

8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.
No additional verbose logging necessary.

9. Anything else do we need to know?
- `kube-proxy` is proxying from what we can tell, its setting up the expected IP tables rules at the node level
- This problem seems to be present across multiple kops versions and multiple kubernetes versions - 1.7.x, 1.8.x
If we feel like this is an upstream kubernetes bug I'm happy to move this but wanted to check here first.

**Comment by boube** [comment]
Hi, the same here :(

**Comment by harsha-y** [comment]
@boube Are you running CoreOS/Container-Linux(we are)? What version of kops+kubernetes versions did you use in your tests assuming you're running the cluster on AWS?

**Comment by boube** [comment]
Yes, using CoreOS stable on AWS. Versions: 

$ kops version
Version 1.8.0
$ kubectl version | grep Server
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.6", GitCommit:"6260bb08c46c31eea6cb538b34a9ceb3e406689c", GitTreeState:"clean", BuildDate:"2017-12-21T06:23:29Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}

**Comment by boube** [comment]
@harsha-y any updates?

**Comment by chrislovecnm** [comment]
This might be a duplicate, as there is an issue with flannel and coreos. Possible dupicate https://github.com/kubernetes/kops/issues/4391

@boube there is a workaround listed in #4391, but I would recommend seeing if flannel has addressed this problem, and we can get flannel updated in kops.

**Comment by boube** [comment]
Hi! It looks like a duplicate. It isn't the exact same behavior but it could be related. 

Thank you very much, feel free to mark it as dup and close this. 
I'm out of the office today, but I will try again this week with and without workaround. If both solutions fail, I will reopen again.

**Comment by harsha-y** [comment]
@chrislovecnm Thanks for the tip! We've ruled out DNS because name resolution never fails within the cluster but #4391 looks like very similar to what we are seeing. Will the give the workaround a shot and report back if that solves the problem.

**Comment by harsha-y** [comment]
@chrislovecnm @boube That totally worked!

We enabled the `br_netfilter` module on all CoreOS masters and nodes. That solved our issue. We have some monitoring around timeouts within our cluster and haven't seen any alerts since we made this change. 😄 

Interesting to note that before we made changes we took inventory of where this module was already present and roughly 30% of our nodes had this module already. We are using the same version of CoreOS across our cluster.
